const bcrypt = require('bcrypt');
const User = require('../models/User');

exports.createUser = async (req, res) => {
  try {
    const { username, password } = req.body;

    const existingUser = await User.findOne({ username });

    if (existingUser) {
      return res.status(400).json({ error: 'El nombre de usuario ya est치 en uso.' });
    }

    const saltRounds = 10;
    const salt = await bcrypt.genSalt(saltRounds);

    const hashedPassword = await bcrypt.hash(password, salt);

    const newUser = new User({ username, password: hashedPassword });
    await newUser.save();

    res.status(201).json({ message: 'Usuario creado exitosamente.' });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Error en el servidor.' });
  }
};

exports.authenticateUser = async (req, res) => {
  try {
    const { username, password } = req.body;
    console.log('Intento de autenticaci칩n para el usuario:', username);

    const user = await User.findOne({ username });

    if (!user) {
      console.log('Usuario no encontrado:', username);
      return res.status(404).json({ error: 'Usuario no encontrado.' });
    }

    const isPasswordValid = await bcrypt.compare(password, user.password);

    if (isPasswordValid) {
      console.log('Autenticaci칩n exitosa para el usuario:', username);
      res.json({ message: 'Autenticaci칩n exitosa', user: { username: user.username, victories: user.victories } });
    } else {
      console.log('Credenciales incorrectas para el usuario:', username);
      res.status(401).json({ error: 'Credenciales incorrectas.' });
    }
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Error en el servidor.' });
  }
};



exports.addVictory = async (req, res) => {
  try {
    const { username } = req.params;

    const user = await User.findOne({ username });

    if (!user) {
      return res.status(404).json({ error: 'Usuario no encontrado.' });
    }

    user.victories += 1;

    await user.save();

    res.json({ victories: user.victories });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Error en el servidor.' });
  }
};
